File: app/src/main/java/com/peterlaurence/trekadvisor/util/gpx/model/TrackPoint.java
package com.peterlaurence.trekadvisor.util.gpx.model;

import java.util.Date;

/**
 * Represents a waypoint, point of interest, or named feature on a map.
 *
 * @author peterLaurence on 12/02/17.
 */
public class TrackPoint {
    private final Double mLatitude;
    private final Double mLongitude;
    private final Double mElevation;
    private final Date mTime;

    private TrackPoint(Builder builder) {
        mLatitude = builder.mLatitude;
        mLongitude = builder.mLongitude;
        mElevation = builder.mElevation;
        mTime = builder.mTime;
    }

    public Double getElevation() {
        return mElevation;
    }

    public Double getLatitude() {
        return mLatitude;
    }

    public Double getLongitude() {
        return mLongitude;
    }

    public Date getTime() {
        return mTime;
    }

    public static class Builder {
        private Double mLatitude;
        private Double mLongitude;
        private Double mElevation;
        private Date mTime;

        public Builder setLatitude(Double latitude) {
            mLatitude = latitude;
            return this;
        }

        public Builder setLongitude(Double longitude) {
            mLongitude = longitude;
            return this;
        }

        public Builder setElevation(Double elevation) {
            mElevation = elevation;
            return this;
        }

        public Builder setTime(Date time) {
            mTime = time;
            return this;
        }

        public TrackPoint build() {
            return new TrackPoint(this);
        }
    }

}

File: app/src/main/java/com/peterlaurence/trekadvisor/util/gpx/model/TrackPoint.kt
package com.peterlaurence.trekadvisor.util.gpx.model

import java.util.Date

/**
 * Represents a waypoint, point of interest, or named feature on a map.
 *
 * @author peterLaurence on 12/02/17.
 */
class TrackPoint(
        val latitude: Double,
        val longitude: Double,
        val elevation: Double?,
        val time: Date?
) {
    private constructor(builder: Builder) : this(builder.latitude, builder.longitude,
            builder.elevation, builder.time)

    class Builder {
        var latitude: Double = 0.0
            private set
        var longitude: Double = 0.0
            private set
        var elevation: Double? = null
            private set
        var time: Date? = null
            private set

        fun setLatitude(latitude: Double) = apply { this.latitude = latitude }

        fun setLongitude(longitude: Double) = apply { this.longitude = longitude }

        fun setElevation(elevation: Double?) = apply { this.elevation = elevation }

        fun setTime(time: Date) = apply { this.time = time }

        fun build() = TrackPoint(this)
    }
}
